oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java

```diff
@@ -0,0 +1,357 @@
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import org.heigit.ohsome.oshdb.osh.OSHEntity;
+import org.heigit.ohsome.oshdb.osh.OSHNode;
+import org.heigit.ohsome.oshdb.osh.OSHRelation;
+import org.heigit.ohsome.oshdb.osh.OSHWay;
+import org.heigit.ohsome.oshdb.osm.OSMEntity;
+import org.heigit.ohsome.oshdb.osm.OSMMember;
+import org.heigit.ohsome.oshdb.osm.OSMNode;
+import org.heigit.ohsome.oshdb.osm.OSMRelation;
+import org.heigit.ohsome.oshdb.osm.OSMWay;
+import org.heigit.ohsome.oshdb.util.OSHDBTimestamp;
+
+public class OSHEntityTimeUtils {
+  private OSHEntityTimeUtils() {
+    throw new IllegalStateException("utility class");
+  }
+
+  /**
+   * Returns the changeset ids which correspond to modifications of this entity.
+   *
+   * <p>Used internally to group modifications by changeset.</p>
+   *
+   * @param osh the osh entity to work on
+   * @return a map between timestamps and changeset ids
+   */
+  public static Map<OSHDBTimestamp, Long> getChangesetTimestamps(OSHEntity osh) {
+    switch (osh.getType()) {
+      case NODE:
+        return getChangesetTimestamps((OSHNode) osh);
+      case WAY:
+        return getChangesetTimestamps((OSHWay) osh);
+      case RELATION:
+        return getChangesetTimestamps((OSHRelation) osh);
+      default:
+        throw new UnsupportedOperationException(format(UNSUPPORTED_OSMTYPE_MESSAGE, osh.getType()));
+    }
+  }
+
+  private static Map<OSHDBTimestamp, Long> getChangesetTimestamps(OSHNode osh) {
+    Map<OSHDBTimestamp, Long> result = new TreeMap<>();
+    osh.getVersions().forEach(osm -> result.put(osm.getTimestamp(), osm.getChangesetId()));
+    return result;
+  }
+
+  private static Map<OSHDBTimestamp, Long> getChangesetTimestamps(OSHWay osh) {
+    Map<OSHDBTimestamp, Long> result = new TreeMap<>();
+
+    List<OSMWay> ways = org.heigit.ohsome.oshdb.osh.OSHEntities.toList(osh.getVersions());
+    ways.forEach(osm -> result.put(osm.getTimestamp(), osm.getChangesetId()));
+
+    // recurse way nodes
+    try {
+      for (OSHNode node : osh.getNodes()) {
+        if (node == null) {
+          continue;
+        }
+        for (OSMNode version : node.getVersions()) {
+          result.putIfAbsent(version.getTimestamp(), version.getChangesetId());
+        }
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    return result;
+  }
+
+  private static Map<OSHDBTimestamp, Long> getChangesetTimestamps(OSHRelation osh) {
+    Map<OSHDBTimestamp, Long> result = new TreeMap<>();
+
+    List<OSMRelation> rels = org.heigit.ohsome.oshdb.osh.OSHEntities.toList(osh.getVersions());
+    rels.forEach(osmRel -> result.put(osmRel.getTimestamp(), osmRel.getChangesetId()));
+
+    // recurse rel members
+    try {
+      Stream.concat(osh.getNodes().stream(), osh.getWays().stream())
+          .filter(Objects::nonNull)
+          .forEach(oshEntity -> getChangesetTimestamps(oshEntity).forEach(result::putIfAbsent));
```

Comment:
in `getChangesetTimestamps(OSHWay osh)` we use a for-each loop instead of stream for the members.
To make this more consistent, me suggestion would be:
```java
for (var mem : osh.getNodes()) {
  if (mem == null) {
    continue;
  }
  for (var version : mem.getVersions()) {
    result.putIfAbsent(version.getTimestamp(), version.getChangesetId());
  }
}

for (var mem : osh.getWays()) {
  if (mem == null) {
    continue;
  }
  getChangesetTimestamps(mem).forEach(result::putIfAbsent);
}
```

Reply:
I didn't like the duplication too much when done exactly this way.

Instead, I will replace it with a new helper method which can be shared between `getChangesetTimestamps(OSHWay)` and `getChangesetTimestamps(OSHRelation)`.

Reply:
ah, I forgot/overlooked that the second loop need to call `getChangesetTimestamps`. :see_no_evil: